Set version early enough not to be overridden by winfx targets on desktop

nguerrera opened the pull request:
**Customer scenario**

* net4x package version should be same as other platforms (1.0.0), but it is currently 1.0.0.0 (one too many zeros)
* VersionPrefix has no impact on net4x projects
* Projects multi-targeting net4x and netcoreapp and using default version will fail to nuget restore in VS

Fix #814
Fix #422 
Fix #93 


**Workarounds, if any**

None for VersionPrefix not working on net4x.
For the others, specify an explicit Version in your .csproj


**Risk**

Low. 

**Performance impact**

None 

**Is this a regression from a previous update?**

The nuget restore may have worked before but the version prefix issue has always existed. The fix is the same for both.

**Root cause analysis:**

Missing test coverage for version prefix using desktop msbuild (added with PR).

**How was the bug found?**

Dogfooding and customer reported

@srivatsn @dsplaisted @eerhardt @dotnet/project-system

nguerrera committed:
Move assembly/package info defaults to before common targets

This prevents a conflict with Version being set to 1.0.0.0 by
desktop targets before we have a chance to set it to 1.0.0, which
was breaking restore. It also breaks VersionPrefix for net4x
projects built with desktop msbuild.

Also leaving VersionPrefix unset if we do not use it.

nguerrera committed:
Add test for version prefix case

srivatsn commented:
Tagging @MattGertz for approval.

eerhardt commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToControlGeneratedAssemblyInfo.cs:
```diff
@@ -82,5 +83,34 @@ public void It_respects_opt_outs(string attributeToOptOut)
+            }
+
+            var testAsset = _testAssetsManager
+                .CopyTestAsset("HelloWorld", identifier: targetFramework)
```
Does the identifier need to be unique by itself?  Or is something else appended to it to make it unique.

If it has to be unique, just the TFM probably isn't enough.

nguerrera replied:
It is appended to the caller member name.

nil4 commented on src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.BeforeCommon.targets:
```diff
@@ -15,6 +15,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
   </PropertyGroup>
 
+  <Import Project="$(MSBuildThisFileDriectory)Microsoft.NET.DefaultAssemblyInfo.targets" />
```
typo: `Driectory`

nguerrera replied:
Good catch. I guess imports are already relative to current dir, this is unnecessary and evaluating to empty, which still works. I'll correct the typo and keep it this way for consistency but we should probably remove the unnecessary uses of MSBuildThisFileDirectory for readability at some point.

Petermarcu commented:
@dotnet-bot test this please

nguerrera committed:
Fix benign typo

This was benign because imports are already relative to the importing
project and the misspelling of MSBuildThisFileDirectory evaluated to
empty.

srivatsn commented:
Approved offline
